clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -208,16 +208,28 @@ static bool isZeroSizedArray(QualType Ty) {
 
 static Sema::DeviceDiagBuilder
 emitDeferredDiagnosticAndNote(Sema &S, SourceRange Loc, unsigned DiagID,
-                              SourceRange UsedAtLoc) {
-  Sema::DeviceDiagBuilder builder =
-      S.SYCLDiagIfDeviceCode(Loc.getBegin(), DiagID);
-  if (UsedAtLoc.isValid())
-    S.SYCLDiagIfDeviceCode(UsedAtLoc.getBegin(), diag::note_sycl_used_here);
+                              bool DiagImmediately, SourceRange UsedAtLoc) {
+  FunctionDecl *FD = dyn_cast<FunctionDecl>(S.getCurLexicalContext());
+  auto builder =
+      DiagImmediately
```

Comment:
Flag for immediate diagnostic doesn't seem right. AFAIK deferred diagnostics are emitted immediately if we know that the code will be emitted for device. I think we need to modify `getEmissionStatus` function for SYCL, so it will correctly identify generated OpenCL kernel as device code.

Reply:
There are two things going on here.  First, the function returns a `Sema::DeviceDiagBuilder` so that the caller can continue to pipe info to the diagnostic  ( ` builder <<  Ty`  ).      `Sema::DeviceDiagBuilder` does not have `DiagEngine` as its base class. So, if an immediate diagnostic is wanted, we have to use one of the DeviceDiagBuilder constructors . This use case (of behaving like Sema::Diag()) is specifically mentioned in the class declaration in Sema.h.  That's why I'm using it.  I can make a nicer wrapper so it's less ugly. 

The second is that, AFAICT, we need an immediate diagnostic. The other diagnostics in  `SyclKernelFieldCheckeer` all use direct diagnostics (`Sema::Diag`). And, that makes sense, this visitor is for the kernel, there is no question about that.   
@Fznamznon you are right that the deferred diagnostic function checks emission status to simply escalate a deferred diag to immediate if the context function has already been emitted. But, the current context doesn't seem anymore to be the original function - the current function context has a munged name and does not have either the `SYCLDeviceAttr` or `SYCLKernelAttr` .  So it's emission status ends up Unknown, not Emitted.

If you two have any suggestions, let me know.    


Working on another part, I have tripped on a "Multiple diagnostics in flight at once!" assertion.  So the diagnostic part of this might be changing anyway.

Reply:
We have a flag that indicates that we are constructing OpenCL kernel. This is used to don't emit deferred diagnostics at all. See https://github.com/intel/llvm/blob/714642633b59bf3642438ec910e39493fefae750/clang/lib/Sema/SemaSYCL.cpp#L1543
I don't recall the exact reason why we added it, but it seems that removing this flag from the check which I listed above doesn't break anything in SemaSYCL tests. So, I thinks you can try to use it and add a case in `SYCLDiagIfDeviceCode` that returns `DeviceDiagBuilder::K_ImmediateWithCallStack` if we are constucting OpenCL kernel

Reply:
I was wondering about that flag. Thanks!

Reply:
I've taken @Fznamznon  suggestion and now the `ConstructingOpenCLKernel` flag, when true means a deferred diagnostic will be immediate.  This simplifies things.